File: modules/agent/src/controllers/organization_controller.py
"""Controller layer for organization routing to services."""

from typing import List, Optional

from fastapi import Request

from lib.decorators import handle_http_exceptions
from serializers.common import to_paged_response, contact_to_dict
from serializers.organization import (
    technology_to_dict,
    funding_round_to_dict,
    signal_to_dict,
    org_to_list_response,
    org_to_search_response,
    org_to_detail_response,
)
from services.organization_service import (
    OrganizationCreate,
    OrganizationUpdate,
    OrgContactCreate,
    OrgContactUpdate,
    OrgTechnologyCreate,
    FundingRoundCreate,
    SignalCreate,
    get_organizations_paginated,
    search_organizations as search_organizations_service,
    get_organization as get_organization_service,
    create_organization as create_organization_service,
    update_organization as update_organization_service,
    delete_organization as delete_organization_service,
    get_organization_contacts as get_contacts_service,
    create_organization_contact as create_contact_service,
    update_organization_contact as update_contact_service,
    delete_organization_contact as delete_contact_service,
    get_technologies as get_technologies_service,
    add_technology as add_technology_service,
    remove_technology as remove_technology_service,
    get_funding_rounds as get_funding_rounds_service,
    create_funding_round as create_funding_round_service,
    delete_funding_round as delete_funding_round_service,
    get_signals as get_signals_service,
    create_signal as create_signal_service,
    delete_signal as delete_signal_service,
)


# Re-export for routes
__all__ = [
    "OrganizationCreate",
    "OrganizationUpdate",
    "OrgContactCreate",
    "OrgContactUpdate",
    "OrgTechnologyCreate",
    "FundingRoundCreate",
    "SignalCreate",
]


# Organization CRUD

@handle_http_exceptions
async def get_organizations(
    request: Request,
    page: int,
    limit: int,
    order_by: str,
    order: str,
    search: Optional[str] = None,
) -> dict:
    """Get organizations with pagination."""
    tenant_id = request.state.tenant_id
    organizations, total = await get_organizations_paginated(
        tenant_id, page, limit, order_by, order, search
    )
    items = [org_to_list_response(org) for org in organizations]
    return to_paged_response(total, page, limit, items)


@handle_http_exceptions
async def search_organizations(request: Request, query: str) -> List[dict]:
    """Search organizations by name."""
    tenant_id = request.state.tenant_id
    organizations = await search_organizations_service(query, tenant_id)
    return [org_to_search_response(org) for org in organizations]


@handle_http_exceptions
async def get_organization(org_id: int) -> dict:
    """Get organization by ID with contacts and research data."""
    org, contacts = await get_organization_service(org_id)
    return org_to_detail_response(org, contacts=contacts, include_research=True)


@handle_http_exceptions
async def create_organization(request: Request, data: OrganizationCreate) -> dict:
    """Create a new organization."""
    tenant_id = request.state.tenant_id
    user_id = request.state.user_id
    org_data = data.model_dump(exclude_none=True)
    industry_ids = org_data.pop("industry_ids", None)
    project_ids = org_data.pop("project_ids", None)
    org_data["created_by"] = user_id
    org_data["updated_by"] = user_id
    org = await create_organization_service(org_data, tenant_id, industry_ids, project_ids)
    return org_to_detail_response(org)


@handle_http_exceptions
async def update_organization(request: Request, org_id: int, data: OrganizationUpdate) -> dict:
    """Update an organization."""
    user_id = request.state.user_id
    org_data = data.model_dump(exclude_unset=True)
    industry_ids = org_data.pop("industry_ids", None)
    project_ids = org_data.pop("project_ids", None)
    project_ids_provided = "project_ids" in data.model_dump(exclude_unset=True)
    org_data["updated_by"] = user_id
    org = await update_organization_service(org_id, org_data, industry_ids, project_ids, project_ids_provided)
    return org_to_detail_response(org)


@handle_http_exceptions
async def delete_organization(org_id: int) -> dict:
    """Delete an organization."""
    await delete_organization_service(org_id)
    return {"success": True}


# Contact management

@handle_http_exceptions
async def get_organization_contacts(org_id: int) -> List[dict]:
    """Get contacts for an organization."""
    contacts = await get_contacts_service(org_id)
    return [contact_to_dict(c) for c in contacts]


@handle_http_exceptions
async def create_organization_contact(request: Request, org_id: int, data: OrgContactCreate) -> dict:
    """Create a contact for an organization."""
    user_id = request.state.user_id
    contact = await create_contact_service(org_id, data.model_dump(), user_id)
    return contact_to_dict(contact)


@handle_http_exceptions
async def update_organization_contact(
    request: Request,
    org_id: int,
    contact_id: int,
    data: OrgContactUpdate,
) -> dict:
    """Update a contact for an organization."""
    user_id = request.state.user_id
    contact = await update_contact_service(org_id, contact_id, data.model_dump(exclude_unset=True), user_id)
    return contact_to_dict(contact)


@handle_http_exceptions
async def delete_organization_contact(org_id: int, contact_id: int) -> dict:
    """Delete a contact from an organization."""
    await delete_contact_service(org_id, contact_id)
    return {"success": True}


# Technology management

@handle_http_exceptions
async def get_organization_technologies(org_id: int) -> dict:
    """Get technologies for an organization."""
    technologies = await get_technologies_service(org_id)
    return {"technologies": [technology_to_dict(t) for t in technologies]}


@handle_http_exceptions
async def add_organization_technology(
    org_id: int,
    tech_id: int,
    source: Optional[str],
    confidence: Optional[float],
) -> dict:
    """Add a technology to an organization."""
    org_tech = await add_technology_service(org_id, tech_id, source, confidence)
    return {"organization_technology": technology_to_dict(org_tech)}


@handle_http_exceptions
async def remove_organization_technology(org_id: int, technology_id: int) -> dict:
    """Remove a technology from an organization."""
    await remove_technology_service(org_id, technology_id)
    return {"success": True}


# Funding round management

@handle_http_exceptions
async def get_organization_funding_rounds(org_id: int) -> dict:
    """Get funding rounds for an organization."""
    funding_rounds = await get_funding_rounds_service(org_id)
    return {"funding_rounds": [funding_round_to_dict(fr) for fr in funding_rounds]}


@handle_http_exceptions
async def create_organization_funding_round(org_id: int, round_data: Dict[str, Any]) -> dict:
    """Create a funding round for an organization."""
    funding_round = await create_funding_round_service(org_id, round_data)
    return {"funding_round": funding_round_to_dict(funding_round)}


@handle_http_exceptions
async def delete_organization_funding_round(org_id: int, round_id: int) -> dict:
    """Delete a funding round."""
    await delete_funding_round_service(org_id, round_id)
    return {"success": True}


# Signal management

@handle_http_exceptions
async def get_organization_signals(org_id: int, signal_type: Optional[str], limit: int) -> dict:
    """Get signals for an organization."""
    signals = await get_signals_service(org_id, signal_type, limit)
    return {"signals": [signal_to_dict(s) for s in signals]}


@handle_http_exceptions
async def create_organization_signal(org_id: int, signal_data: Dict[str, Any]) -> dict:
    """Create a signal for an organization."""
    signal = await create_signal_service(org_id, signal_data)
    return {"signal": signal_to_dict(signal)}


@handle_http_exceptions
async def delete_organization_signal(org_id: int, signal_id: int) -> dict:
    """Delete a signal."""
    await delete_signal_service(org_id, signal_id)
    return {"success": True}

File: modules/agent/internal/controllers/organization_controller.go
package controllers

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"

	"agent/internal/lib"
	"agent/internal/middleware"
	"agent/internal/schemas"
	"agent/internal/serializers"
	"agent/internal/services"
)

// OrganizationController handles organization HTTP requests
type OrganizationController struct {
	orgService *services.OrganizationService
}

// NewOrganizationController creates a new organization controller
func NewOrganizationController(orgService *services.OrganizationService) *OrganizationController {
	return &OrganizationController{orgService: orgService}
}

// GetOrganizations handles GET /organizations
func (c *OrganizationController) GetOrganizations(w http.ResponseWriter, r *http.Request) {
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	orderBy := r.URL.Query().Get("orderBy")
	order := r.URL.Query().Get("order")
	search := r.URL.Query().Get("search")

	var tenantID *int64
	if tid, ok := middleware.GetTenantID(r.Context()); ok {
		tenantID = &tid
	}

	result, err := c.orgService.GetOrganizations(page, limit, orderBy, order, search, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetOrganization handles GET /organizations/{id}
func (c *OrganizationController) GetOrganization(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	result, err := c.orgService.GetOrganization(id)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// SearchOrganizations handles GET /organizations/search
func (c *OrganizationController) SearchOrganizations(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("q")
	if query == "" {
		writeJSON(w, http.StatusOK, []interface{}{})
		return
	}

	var tenantID *int64
	if tid, ok := middleware.GetTenantID(r.Context()); ok {
		tenantID = &tid
	}

	limit := lib.ParseIntQueryParam(r, "limit", 10)

	results, err := c.orgService.SearchOrganizations(query, tenantID, limit)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, results)
}

// DeleteOrganization handles DELETE /organizations/{id}
func (c *OrganizationController) DeleteOrganization(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	err = c.orgService.DeleteOrganization(id)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, serializers.SuccessResponse{Success: true})
}

// UpdateOrganization handles PUT /organizations/{id}
func (c *OrganizationController) UpdateOrganization(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	var input schemas.OrganizationUpdate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	userID, _ := middleware.GetUserID(r.Context())

	result, err := c.orgService.UpdateOrganization(id, input, userID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// AddOrganizationContact handles POST /organizations/{id}/contacts
func (c *OrganizationController) AddOrganizationContact(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	var input schemas.OrgContactCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	userID, _ := middleware.GetUserID(r.Context())
	tenantID, _ := middleware.GetTenantID(r.Context())

	result, err := c.orgService.AddContactToOrganization(orgID, input, userID, tenantID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, result)
}

// CreateOrganization handles POST /organizations
func (c *OrganizationController) CreateOrganization(w http.ResponseWriter, r *http.Request) {
	var input schemas.OrganizationCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	var tenantID *int64
	if tid, ok := middleware.GetTenantID(r.Context()); ok {
		tenantID = &tid
	}
	userID, _ := middleware.GetUserID(r.Context())

	result, err := c.orgService.CreateOrganization(input, userID, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, result)
}

// GetOrganizationContacts handles GET /organizations/{id}/contacts
func (c *OrganizationController) GetOrganizationContacts(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	result, err := c.orgService.GetOrganizationContacts(orgID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// UpdateOrganizationContact handles PUT /organizations/{id}/contacts/{contactId}
func (c *OrganizationController) UpdateOrganizationContact(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	contactIDStr := chi.URLParam(r, "contactId")
	contactID, err := strconv.ParseInt(contactIDStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid contact ID")
		return
	}

	var input schemas.OrgContactUpdate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	userID, _ := middleware.GetUserID(r.Context())

	result, err := c.orgService.UpdateOrganizationContact(orgID, contactID, input, userID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// DeleteOrganizationContact handles DELETE /organizations/{id}/contacts/{contactId}
func (c *OrganizationController) DeleteOrganizationContact(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	contactIDStr := chi.URLParam(r, "contactId")
	contactID, err := strconv.ParseInt(contactIDStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid contact ID")
		return
	}

	err = c.orgService.DeleteOrganizationContact(orgID, contactID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, serializers.SuccessResponse{Success: true})
}

// GetOrganizationTechnologies handles GET /organizations/{id}/technologies
func (c *OrganizationController) GetOrganizationTechnologies(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	result, err := c.orgService.GetOrganizationTechnologies(orgID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// AddOrganizationTechnology handles POST /organizations/{id}/technologies
func (c *OrganizationController) AddOrganizationTechnology(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	var input schemas.OrgTechnologyCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	err = c.orgService.AddTechnologyToOrganization(orgID, input)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, serializers.SuccessResponse{Success: true})
}

// RemoveOrganizationTechnology handles DELETE /organizations/{id}/technologies/{techId}
func (c *OrganizationController) RemoveOrganizationTechnology(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	techIDStr := chi.URLParam(r, "techId")
	techID, err := strconv.ParseInt(techIDStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid technology ID")
		return
	}

	if err := c.orgService.RemoveTechnologyFromOrganization(orgID, techID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, serializers.SuccessResponse{Success: true})
}

// GetOrganizationFundingRounds handles GET /organizations/{id}/funding-rounds
func (c *OrganizationController) GetOrganizationFundingRounds(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	result, err := c.orgService.GetOrganizationFundingRounds(orgID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// CreateOrganizationFundingRound handles POST /organizations/{id}/funding-rounds
func (c *OrganizationController) CreateOrganizationFundingRound(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	var input schemas.FundingRoundCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	result, err := c.orgService.CreateOrganizationFundingRound(orgID, input)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, result)
}

// DeleteOrganizationFundingRound handles DELETE /organizations/{id}/funding-rounds/{roundId}
func (c *OrganizationController) DeleteOrganizationFundingRound(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	roundIDStr := chi.URLParam(r, "roundId")
	roundID, err := strconv.ParseInt(roundIDStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid round ID")
		return
	}

	if err := c.orgService.DeleteOrganizationFundingRound(orgID, roundID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, serializers.SuccessResponse{Success: true})
}

// GetOrganizationSignals handles GET /organizations/{id}/signals
func (c *OrganizationController) GetOrganizationSignals(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	var signalType *string
	if st := r.URL.Query().Get("signal_type"); st != "" {
		signalType = &st
	}

	limit := lib.ParseIntQueryParam(r, "limit", 20)

	result, err := c.orgService.GetOrganizationSignals(orgID, signalType, limit)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// CreateOrganizationSignal handles POST /organizations/{id}/signals
func (c *OrganizationController) CreateOrganizationSignal(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	var input schemas.SignalCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	userID, _ := middleware.GetUserID(r.Context())

	result, err := c.orgService.CreateOrganizationSignal(orgID, input, userID)
	if errors.Is(err, services.ErrOrganizationNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, result)
}

// DeleteOrganizationSignal handles DELETE /organizations/{id}/signals/{signalId}
func (c *OrganizationController) DeleteOrganizationSignal(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	orgID, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid organization ID")
		return
	}

	signalIDStr := chi.URLParam(r, "signalId")
	signalID, err := strconv.ParseInt(signalIDStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid signal ID")
		return
	}

	if err := c.orgService.DeleteOrganizationSignal(orgID, signalID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, serializers.SuccessResponse{Success: true})
}
